keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+
+- [Kubernetes Component Configuration](https://docs.google.com/document/d/1arP4T9Qkp2SovlJZ_y790sBeiWXDO6SG10pZ_UUU-Lc/edit) by [@mikedanese](https://github.com/mikedanese)
+- [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit#) by [@mtaufen](https://github.com/mtaufen)
+- [Moving ComponentConfig API types to staging repos](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md) by [@luxas](https://github.com/luxas) and [@sttts](https://github.com/sttts)
+
+## Proposal
+
+This proposal contains three logical units of work. Each subsection is explained in more detail below.
+
+### Part 1: ComponentConfig
+
+#### Standardized encoding/decoding
+
+- Encoding/decoding helper methods in `k8s.io/component/config/serializer` that would be referenced in every scheme package
+- Warn or (if desired, error) on unknown fields
+  - This has the benefit that it makes it possible for the user to spot e.g. config typos. More high-level, this can be used for e.g. a `--validate-config` flag
+- Support both JSON and YAML
+- Support multiple YAML documents if needed
+
+#### Testing helper methods
+
+- Conversion / roundtrip testing
+- API group testing
+  - External types must have JSON tags
+  - Internal types must not have any JSON tags
+  - ...
+- Defaulting testing
+
+#### Generate OpenAPI specifications
+
+Provide a common way to generate OpenAPI specifications local to the component, so that external consumers can access it, and the component can expose it via e.g. a CLI flag or HTTPS endpoint.
+
+### Part 2: Command building / flag parsing
+
+#### Wrapper around cobra.Command
+
+See the `cmd/kubelet` code for how much extra setup a Kubernetes component needs to do for building commands and flag sets. This code can be refactored into a generic wrapper around _cobra_ for use with Kubernetes.
+
+#### Flag precedence over config file
+
+If the component supports both ComponentConfiguration and flags, flags should override fields set in the ComponentConfiguration. This is not straightforward to implement in code, and only the kubelet does this at the moment. Refactoring this code in a generic helper library in this new repository will make adoption of the feature easy and testable. The details of flag versus ComponentConfig semantics are to be decided later in a different proposal. Meanwhile, this flag precedence feature will be opt-in, so the kubelet and kubeadm can directly adopt this code, until the details have been decided on for all components.
+
+#### Standardized logging
+
+Use the _k8s.io/klog_ package in a standardized way.
+
+### Part 3: HTTPS serving
+
+Many Kubernetes controllers are clients to the API server and run as daemons. In order to expose information on how the component is doing (e.g. profiling, metrics, current configuration, etc.), an HTTPS server is run.
```

Comment:
Also streaming things like logs, exec, attach.

Reply:
Why logs, exec, attach? These are very special API server endpoints. They don't even belong into the generic API server code.

Reply:
For any frameworks we build, we should at least ensure that users don't have to make a binary choice between using the framework and rolling everything on their own when they want something custom. Calling out uncommon things helps us keep that in mind.

Reply:
ok, I'll add this as a non-goal here. The addition of these packages don't exclude any streaming features though.